src/lib/sql/sql_pipeline_statement.cpp

```diff
@@ -10,31 +10,33 @@
 #include "logical_query_plan/lqp_translator.hpp"
 #include "optimizer/optimizer.hpp"
 #include "scheduler/current_scheduler.hpp"
+#include "sql/sql_pipeline_control_block.hpp"
 #include "sql/sql_translator.hpp"
 #include "utils/assert.hpp"
 
 namespace opossum {
 
-SQLPipelineStatement::SQLPipelineStatement(const std::string& sql, bool use_mvcc)
-    : _sql_string(sql), _use_mvcc(use_mvcc), _auto_commit(_use_mvcc) {}
+SQLPipelineStatement::SQLPipelineStatement(const std::string& sql, UseMvcc use_mvcc)
+    : SQLPipelineStatement(
+          sql, std::make_shared<SQLPipelineControlBlock>(use_mvcc, nullptr, Optimizer::get_default_optimizer())) {}
 
 SQLPipelineStatement::SQLPipelineStatement(const std::string& sql,
                                            std::shared_ptr<TransactionContext> transaction_context)
-    : _sql_string(sql), _use_mvcc(true), _auto_commit(false), _transaction_context(std::move(transaction_context)) {
+    : SQLPipelineStatement(sql, std::make_shared<SQLPipelineControlBlock>(UseMvcc::Yes, transaction_context,
+                                                                          Optimizer::get_default_optimizer())) {
   DebugAssert(_transaction_context != nullptr, "Cannot pass nullptr as explicit transaction context.");
 }
 
+SQLPipelineStatement::SQLPipelineStatement(const std::string& sql,
+                                           const std::shared_ptr<const SQLPipelineControlBlock>& control_block)
```

Comment:
Not sure if this is against our guidelines, but why pass in a `shared_ptr&` if you copy it straight away. Wouldn't it be nicer to pass in a value and move it. So if the calling context doesn't need the `SQLPipelineControlBlock` any more, it can just be moved into the constructor and we don't have to deal with ref counting at all. (just an idea, not sure what guidelines say)

Reply:
I KNEW someone would bring this up. std::move(some_shared_ptr), or std::move in general can be quite the cancer simply for reason that a variable suddenly loses its value/goes into some weird moved-out state but still lives on for a while in its scope. This is counter-intuitive because in C++ you generally expect things to live and die with their scope. In a debugger you suddenly see a nullptr that once wasn't a nullptr but it takes time and digging until you realize that it was moved out. std::move makes sense in hot loops, and for vectors, but please let us not create unnecessary problems for ourselves by moving shared ptrs around in code that is absolutely not performance critical.